### ISSUE TITLE
Avoid unused variables

### ISSUE THREAD

**Issue by mcarton** [DotNetAnalyzers/StyleCopAnalyzers#3498]
There should be a rule to prevent unused variables, eg.

```cs
int Method()
{
    var foo = bar(); // warning: variable `foo` is unused
    return baz();
}
```

The codefix would be

```cs
int Method()
{
    bar(); // warning: variable `foo` is unused
    return baz();
}
```

---

This is inspired by a bug in a codebase I'm working on which was caused by the wrong variable being used. The proper variable wasn't used anywhere else, so a rule like this would have prevented this bug. In general, there should be no use for naming a variable and not use it. `using var foo = bar();` is an exception to this: This one must have a name even if it's not used.

**Comment by manfred-brands** [comment]
Was this not reported by [IDE0059](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0059)?
